File: packages/database/src/repositories/identity.ts
/**
 * Identity Repository
 * 
 * Type-safe database operations for Identity model.
 */

import { prisma, Identity, TrainingStatus, Prisma } from '../index'

export class IdentityRepository {
  /**
   * Create new identity
   */
  static async create(data: {
    userId: string
    name: string
    description?: string
    referencePhotoUrls: string[]
  }): Promise<Identity> {
    return prisma.identity.create({
      data: {
        userId: data.userId,
        name: data.name,
        description: data.description,
        referencePhotoUrls: data.referencePhotoUrls as any,
        referencePhotoCount: data.referencePhotoUrls.length,
        trainingStatus: TrainingStatus.PENDING,
      },
    })
  }

  /**
   * Get identity by ID
   */
  static async findById(id: string): Promise<Identity | null> {
    return prisma.identity.findFirst({
      where: { id, isDeleted: false },
      include: {
        user: {
          select: {
            id: true,
            email: true,
            name: true,
            tier: true,
          },
        },
      },
    })
  }

  /**
   * Get user's identities
   */
  static async findByUserId(userId: string): Promise<Identity[]> {
    return prisma.identity.findMany({
      where: { userId, isDeleted: false },
      orderBy: { createdAt: 'desc' },
    })
  }

  /**
   * Update training status
   */
  static async updateTrainingStatus(
    id: string,
    status: TrainingStatus,
    progress: number,
    error?: string
  ): Promise<Identity> {
    const updateData: Prisma.IdentityUpdateInput = {
      trainingStatus: status,
      trainingProgress: progress,
    }

    if (status === TrainingStatus.TRAINING && progress === 0) {
      updateData.trainingStartedAt = new Date()
    }

    if (status === TrainingStatus.COMPLETED) {
      updateData.trainingCompletedAt = new Date()
      updateData.trainingProgress = 100
    }

    if (status === TrainingStatus.FAILED) {
      updateData.trainingError = error
    }

    return prisma.identity.update({
      where: { id },
      data: updateData,
    })
  }

  /**
   * Save LoRA model info
   */
  static async saveLoraModel(
    id: string,
    data: {
      loraFilePath: string
      loraFileSize: number
      faceEmbedding: number[]
      qualityScore: number
      faceConsistencyScore: number
    }
  ): Promise<Identity> {
    return prisma.identity.update({
      where: { id },
      data: {
        loraFilePath: data.loraFilePath,
        loraFileSize: data.loraFileSize,
        faceEmbedding: data.faceEmbedding as any,
        qualityScore: data.qualityScore,
        faceConsistencyScore: data.faceConsistencyScore,
        trainingStatus: TrainingStatus.COMPLETED,
        trainingCompletedAt: new Date(),
      },
    })
  }

  /**
   * Record consent
   */
  static async recordConsent(
    id: string,
    ipAddress: string,
    userAgent: string
  ): Promise<Identity> {
    return prisma.identity.update({
      where: { id },
      data: {
        consentGiven: true,
        consentTimestamp: new Date(),
        consentIpAddress: ipAddress,
        consentUserAgent: userAgent,
        consentVersion: 'v1.0',
      },
    })
  }

  /**
   * Check if identity is ready for generation
   */
  static async isReady(id: string): Promise<{ ready: boolean; reason?: string }> {
    const identity = await prisma.identity.findUnique({
      where: { id },
      select: {
        trainingStatus: true,
        consentGiven: true,
        loraFilePath: true,
        isDeleted: true,
      },
    })

    if (!identity) {
      return { ready: false, reason: 'Identity not found' }
    }

    if (identity.isDeleted) {
      return { ready: false, reason: 'Identity deleted' }
    }

    if (!identity.consentGiven) {
      return { ready: false, reason: 'Consent not given' }
    }

    if (identity.trainingStatus !== TrainingStatus.COMPLETED) {
      return { ready: false, reason: 'Training not completed' }
    }

    if (!identity.loraFilePath) {
      return { ready: false, reason: 'LoRA model not available' }
    }

    return { ready: true }
  }

  /**
   * Increment usage counter
   */
  static async incrementUsage(id: string): Promise<void> {
    await prisma.identity.update({
      where: { id },
      data: {
        totalGenerations: { increment: 1 },
        lastUsedAt: new Date(),
      },
    })
  }

  /**
   * Soft delete identity
   */
  static async softDelete(id: string): Promise<Identity> {
    return prisma.identity.update({
      where: { id },
      data: {
        isDeleted: true,
        deletedAt: new Date(),
      },
    })
  }

  /**
   * Get training queue position
   */
  static async getQueuePosition(id: string): Promise<number> {
    const identity = await prisma.identity.findUnique({
      where: { id },
      select: { createdAt: true },
    })

    if (!identity) return -1

    const position = await prisma.identity.count({
      where: {
        trainingStatus: TrainingStatus.PENDING,
        createdAt: { lt: identity.createdAt },
      },
    })

    return position + 1
  }
}

File: apps/api/app/models/identity.py
"""Identity model for SQLAlchemy."""
from sqlalchemy import Column, String, Integer, Float, Boolean, DateTime, JSON, ForeignKey, Enum as SQLEnum
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import UUID
from datetime import datetime
import uuid
import enum

from app.core.database import Base


class TrainingStatus(str, enum.Enum):
    """Training status enum"""
    PENDING = "PENDING"
    VALIDATING = "VALIDATING"
    PREPROCESSING = "PREPROCESSING"
    TRAINING = "TRAINING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


class Identity(Base):
    """Identity model for LoRA training."""
    
    __tablename__ = "identities"
    
    # Primary key
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    
    # Foreign key
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    
    # Identity info
    name = Column(String(100), nullable=False)
    description = Column(String(500), nullable=True)
    trigger_word = Column(String(20), default="sks", nullable=False)
    
    # Training data
    reference_photo_urls = Column(JSON, nullable=False)  # Array of S3 URLs
    reference_photo_count = Column(Integer, default=0, nullable=False)
    
    # LoRA & Embedding
    lora_file_path = Column(String, nullable=True)  # S3 URL to .safetensors
    lora_file_size = Column(Integer, nullable=True)  # Bytes
    face_embedding = Column(JSON, nullable=True)  # 512-dim vector for InstantID
    
    # Training status
    training_status = Column(SQLEnum(TrainingStatus), default=TrainingStatus.PENDING, nullable=False, index=True)
    training_progress = Column(Integer, default=0, nullable=False)  # 0-100
    training_started_at = Column(DateTime(timezone=True), nullable=True)
    training_completed_at = Column(DateTime(timezone=True), nullable=True)
    training_error = Column(String, nullable=True)
    training_logs = Column(JSON, nullable=True)
    
    # Quality metrics
    quality_score = Column(Float, nullable=True)  # 0-1
    face_consistency_score = Column(Float, nullable=True)  # 0-1
    
    # Consent
    consent_given = Column(Boolean, default=False, nullable=False)
    consent_timestamp = Column(DateTime(timezone=True), nullable=True)
    consent_ip_address = Column(String(45), nullable=True)
    consent_user_agent = Column(String, nullable=True)
    consent_version = Column(String(20), nullable=True)
    
    # Usage
    total_generations = Column(Integer, default=0, nullable=False)
    last_used_at = Column(DateTime(timezone=True), nullable=True)
    
    # Soft delete
    is_deleted = Column(Boolean, default=False, nullable=False, index=True)
    deleted_at = Column(DateTime(timezone=True), nullable=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), default=datetime.utcnow, nullable=False, index=True)
    updated_at = Column(DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    # user = relationship("User", back_populates="identities")
    # generations = relationship("Generation", back_populates="identity")
